manual/src/refman/patterns.etex

```diff
@@ -121,12 +172,29 @@ of length $n$ whose elements match @pattern_1@ \ldots @pattern_n@,
 respectively. This pattern behaves like
 @pattern_1 "::" \ldots "::" pattern_n "::" "[]"@.
 
+\begin{caml_example}{toplevel}
+let rec destutter = function
+  | [] -> []
+  | [a] -> [a]
+  | a :: b :: t -> if a = b then destutter (b :: t) else a :: destutter (b :: t);;
+\end{caml_example}
+
 \subsubsection*{sss:pat-polyvar}{Polymorphic variant patterns}
 
 The pattern @"`"tag-name pattern_1@ matches all polymorphic variants
 whose tag is equal to @tag-name@, and whose argument matches
 @pattern_1@.
 
+\begin{caml_example}{toplevel}
+let rec split = function
+  | [] -> ([], [])
+  | h :: t ->
+      let ss, gs = split t in
+        match h with
+        | `Sheep _ as s -> (s :: ss, gs)
+        | `Goat _ as g -> (ss, g :: gs);;
```

Comment:
I can never decide if this is good style or not, but the brackets around the tuples here aren't necessary:

```
let rec split = function
   | [] -> [], []
   | h :: t ->
       let ss, gs = split t in
         match h with
         | `Sheep _ as s -> s :: ss, gs
         | `Goat _ as g -> ss, g :: gs ;;
```

(this matches the lack of brackets on the pattern matching end over at `let ss, sg = ...`

Reply:
Per the OCaml style guide https://ocaml.org/learn/tutorials/guidelines.html (and I agree - well, mostly. I can't decide either):

How to write pairs

A tuple is parenthesized and the commas therein (delimiters) are each followed by a space: (1, 2), let triplet = (x, y, z)...

Commonly accepted exceptions:

Definition of the components of a pair: In place of let (x, y) = ..., you can write let x, y = ....
Justification: The point is to define several values simultaneously, not to construct a tuple. Moreover, the pattern is set off nicely between let and =.

Matching several values simultaneously: It's okay to omit parentheses around n-tuples when matching several values simultaneously.

match x, y with
| 1, _ -> ...
| x, 1 -> ...
| x, y -> ...

Justification: The point is to match several values in parallel, not to construct a tuple. Moreover, the expressions being matched are set off by match and with, while the patterns are set off nicely by | and ->.

Reply:
I defer to the superior Chicago Manual of OCaml Style :-)